typing/shape.ml

```diff
@@ -547,18 +555,36 @@ let local_weak_reduce shape =
 
 let dummy_mod = { uid = None; desc = Struct Item.Map.empty }
 
-let of_path ~find_shape ~namespace =
+let of_path ~find_shape ~namespace path =
   let rec aux : Sig_component_kind.t -> Path.t -> t = fun ns -> function
     | Pident id -> find_shape ns id
-    | Pdot (path, name) -> proj (aux Module path) (name, ns)
+    | Pdot (path, name) ->
+        (* We need to handle the following cases:
+          Path of constructor:
+            M.t.C
+          Path of label:
+            M.t.lbl
+          Path of label of inline record:
+            M.t.C.lbl *)
+        let is_capitalized name = String.capitalize_ascii name = name in
+        let namespace : Sig_component_kind.t =
+          match path with
+          | Pident id when is_capitalized (Ident.name id) -> Module
+          | Pident _ -> Type
+          | Pdot (_, name)  when is_capitalized name -> Module
+          | Pdot _ -> Type
+          | Papply _ -> Module
+          | Pextra_ty _ -> Constructor
+        in
+        proj (aux namespace path) (name, ns)
```

Comment:
Well when `path.name` is an inline record label, the `path` actually goes to a Constructor from my point of view.

I am trying to understand the difficulty and the thing is that shapes'  "paths" don't have the same meaning as the compiler's paths:
-  `A.t.C` is a real "shape path" while it's not a valid compiler path. It is built from the "real" path of `t`: `Pdot(path_t, "constructor C")` that is `Pdot(Pdot(Ident A, "type t"), "constructor C")`.
- `A.t.lbl` is also a valid "shape path": `Pdot(Pdot(Ident A, "type t"), "label lbl")`. In that case the label is preceded by a type path from both the POW of the compiler and the shapes.
- `A.t.C.lbl` is a bit more tricky since, from the shapes point of view, `C` is a constructor, but from the compiler point of view `A.t.C` is a valid path, not towards `C`, but towards the implicit type of C's record argument. So the label is preceded by a constructor path from the shapes POW, and by a type path from the POW of the compiler.

When we want the shape of a path like `A.t.C.lbl` we ultimately want to project the label `lbl` from the shape of the constructor `C`. And `C` must be projected, as a constructor form the shape of `t`: 
Shapes: `t:{ (Constructor, "C") : { (Label, "lbl") : Leaf } }`.

So when shape-of-pathing: `Pdot(Pextra_ty (Pdot (Ident A, (Type, "t")), Pcstr_ty "C"), (Label, "lbl"))` we have to consider `extra_ty (..., Pcstr_ty "C")` as a Constructor instead of a Type for the projection to succeed.

All this is a bit confuse, and I guess there is a simpler way to formulate, and implement `shape_of_path`. I will give it a try.

Reply:
I pushed a much simpler version in https://github.com/ocaml/ocaml/pull/12508/commits/a85d79bbf6145140cf6c7cd23c1c374a6ba6e4fe

Reply:
My intuition is that if "shape paths" are different from "paths", we should either use two distinct types, or make sure that all "shape paths" can be expressed in `Path.t` even if the rest of the compiler rarely uses them. (This sounds like something that will be useful independently.) But in the meantime, yes, having a less ugly definition of `shape_of_path` is nice, and I agree that your new implementation is much better than the first one.